src/org/opensolaris/opengrok/configuration/messages/MessageListener.java

```diff
@@ -0,0 +1,497 @@
+        if (messageType == null) {
+            throw new IllegalArgumentException("Cannot remove handler for null message type");
+        }
+        if (handler == null) {
+            throw new IllegalArgumentException("Cannot remove null handler");
+        }
+
+        synchronized (messageHandlers) {
+            List<MessageHandler> handlers = messageHandlers.get(messageType);
+            if (handlers == null || !handlers.contains(handler)) {
+                LOGGER.log(Level.WARNING, "{0} is not registered as a handler", handler);
+                return;
+            }
+            handlers.remove(handler);
+            if (handlers.isEmpty()) {
+                messageHandlers.remove(messageType);
+            }
+        }
+    }
+
+    /**
+     * Try to stop the configuration listener thread
+     */
+    public void stopConfigurationListenerThread() {
+        IOUtils.close(configServerSocket);
+    }
+
+    /**
+     * Start a thread to listen on a socket to receive new messages.
+     * The messages can contain various commands for the webapp, including
+     * upload of new configuration.
+     * @param endpoint The socket address to listen on
+     * @throws IOException if the endpoint is not available
+     */
+    public void start(SocketAddress endpoint) throws IOException {
+        if (isListening()) {
+            throw new IllegalStateException("Cannot start when already listening for messages");
+        }
+        bindToAddress(endpoint);
+        new Thread(() -> {
+            while (!configServerSocket.isClosed()) {
+                acceptConnection();
+            }
+        }, "configurationListener").start();
+    }
+
+    private boolean isListening() {
+        return configServerSocket != null && !configServerSocket.isClosed();
+    }
+
+    private void bindToAddress(final SocketAddress address) throws IOException {
+        configServerSocket = new ServerSocket();
+        configServerSocket.bind(address);
+    }
+
+    private void acceptConnection() {
+        try (Socket s = configServerSocket.accept();
+             MessageReader in = new MessageReader(s.getInputStream());
+             OutputStream output = s.getOutputStream()) {
+
+            LOGGER.log(Level.FINE, "OpenGrok: Got request from {0}", s.getInetAddress().getHostAddress());
+
+            Message m = in.readMessage();
+            handleMessage(m, output);
+
+        } catch (SocketException e) {
+            // if closed then it is okay (socket was closed while blocked in accept())
+            if (!configServerSocket.isClosed()) {
+                LOGGER.log(Level.SEVERE, "Socket error", e);
+            }
+        } catch (IOException e) {
+            LOGGER.log(Level.SEVERE, "Error reading config file: ", e);
+        } catch (RuntimeException e) {
+            LOGGER.log(Level.SEVERE, "Error parsing config file: ", e);
+        } catch (ValidationException e) {
+            LOGGER.log(Level.WARNING, "Received invalid message", e);
```

Comment:
if there is unhandled exception it will cause that opengrok will need to be restarted, isn't it better to check for Exception or Throwable?